### ISSUE TITLE
Ram & Cores Mismatch

### ISSUE THREAD

**Issue by mehedihasanziku** [kaliiiiiiiiii/Selenium-Profiles#58]
**Describe the bug**
Navigator.hardwareConcurrency:  does not match worker scope

**To Reproduce**

same as documentation

**Expected behavior or error-message**

 ram and core should be probably match, unfortunately it fails to match or spoof both

**Environment (please complete the following information):**
 - OS: Windows
- OS-Version 10
- Browser Google-Chrome
 - Selenium-Profiles Version 2.2.6

**Additional context**

![image](https://github.com/kaliiiiiiiiii/Selenium-Profiles/assets/48652123/fbf4e1a9-5715-4ab1-aedb-f0d519961fd8)

**Comment by kaliiiiiiiiii** [comment]
1. RAM and cores don't need to be the same, as RAM is the amount of GB available on your device (4), and cores the amount of cores your processor has (8)
2. "does not match worker scope" is a bug inside chrome (see [crbug#1358491](https://bugs.chromium.org/p/chromium/issues/detail?id=1358491)) . Also, fixing this is allready in the [TODO's](https://github.com/kaliiiiiiiiii/Selenium-Profiles#todo)
3. Can you show me the error, when you click on the link at "lies(1)" ?
![image](https://github.com/kaliiiiiiiiii/Selenium-Profiles/assets/89038706/e0a41d4a-9218-47b7-b442-7ab715a13389)

**Comment by mehedihasanziku** [comment]
> 3. Can you show me the error, when you click on the link at "lies(1)" ?

![image](https://github.com/kaliiiiiiiiii/Selenium-Profiles/assets/48652123/378a5791-d1b5-4e43-a850-2fbc4fc7c503)

It shows that error if `RAM and cores` mismatch, but technically both can be different.

**Comment by kaliiiiiiiiii** [comment]
> > 3. Can you show me the error, when you click on the link at "lies(1)" ?
> >    ![image](https://user-images.githubusercontent.com/48652123/244881919-378a5791-d1b5-4e43-a850-2fbc4fc7c503.png)
> 
> @kaliiiiiiiiii It shows that error if `RAM and cores` mismatch, but technically both can be different.

Oh now I think I got what you mean.

- Like for example if you emulate a different number of cores than your device actually has?
- Why do you think it has something to do with the RAM? With javascript, you'd get the value with `navigator.deviceMemory`, but there's not option to emulate that yet ([crbug#1308873](https://bugs.chromium.org/p/chromium/issues/detail?id=1308873)

**Comment by mehedihasanziku** [comment]
> * Like for example if you emulate a different number of cores than your device actually has?

Yes,
I have now cores-4 ram-4
After `Selenium-Profiles` cores-8 ram-4



> * Why do you think it has something to do with the RAM? With javascript, you'd get the value with `navigator.deviceMemory`, but there's not option to emulate that yet ([crbug#1308873](https://bugs.chromium.org/p/chromium/issues/detail?id=1308873)


It's just an thought, if i remove cores, no lies found
Yes, no way, But I searched that,  not sure how to implement

```
https://stackoverflow.com/questions/59367665/spoof-navigator-properties-in-chrome-selenium-webdriver
driver.execute_cdp_cmd("Page.addScriptToEvaluateOnNewDocument", {
  "source": """
    Object.defineProperty(navigator, 'deviceMemory', {
      get: () => 99
    })
  """
})
```

**Comment by kaliiiiiiiiii** [comment]
> ```python
> https://stackoverflow.com/questions/59367665/spoof-navigator-properties-in-chrome-selenium-webdriver
> driver.execute_cdp_cmd("Page.addScriptToEvaluateOnNewDocument", {
>   "source": """
>     Object.defineProperty(navigator, 'deviceMemory', {
>       get: () => 99
>     })
>   """
> })
> ```

Yes, this is indeed an option, but `creep-js` is gonna detect it. Currently, I don't have a way to overwrite javascript values without the possibility of detection.

https://github.com/abrahamjuliot/creepjs/issues/147 or [adtechmadness.wordpress.com/2019/03/23/javascript-tampering-detection-and-stealth](https://adtechmadness.wordpress.com/2019/03/23/javascript-tampering-detection-and-stealth/ ) might give you a starting point why overwriting isn't that easy

**Comment by kaliiiiiiiiii** [comment]
> It's just an thought, if i remove cores, no lies found
yes, indeed. The emulation causes some other leaks as well inside worker scopes:
- Navigator.userAgent
- Navigator.platform
- navigator.hardwareConcurrency

**Comment by kaliiiiiiiiii** [comment]
related to https://github.com/kaliiiiiiiiii/Selenium-Profiles/discussions/50

**Comment by kaliiiiiiiiii** [comment]
might be possible to implement using something like
```js
chrome.webRequest.onBeforeRequest.addListener(
  function(details) {
    // apply profile here
    chrome.tabs.executeScript(details['tabId'],{"code":"debugger"},console.log)
    console.log(details);
    return {};
  },
  {urls: ["<all_urls>"]},
  ["blocking"]
);
```

in selenium-injector

**Comment by Fusseldieb** [comment]
Yep, this issue still persists

**Comment by MarioCodarin** [comment]
I solved matching the cpu cores of my pc with the cores in the profile.

`from selenium_profiles.webdriver import Chrome
from selenium_profiles.profiles import profiles
from selenium.webdriver.common.by import By  # locate elements
from seleniumwire import webdriver
import multiprocessing


profile = profiles.Windows() # or .Android
profile['cdp']['cores'] = multiprocessing.cpu_count()
options = webdriver.ChromeOptions()
options.add_argument("--headless=new")
driver = Chrome(profile, options=options,
                uc_driver=False
                )`

**Comment by mehedihasanziku** [comment]
> I solved matching the cpu cores of my pc with the cores in the profile.

It’s already discussed above. we are looking for a way to bypass without match. because it’s easily fingerprinted.
